presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -0,0 +1,1970 @@
+                for (Map.Entry<List<String>, Action<PartitionAndMore>> partitionEntry : tableEntry.getValue().entrySet()) {
+                    List<String> partitionValues = partitionEntry.getKey();
+                    Action<PartitionAndMore> action = partitionEntry.getValue();
+                    switch (action.getType()) {
+                        case DROP:
+                            metastoreDeleteOperations.add(new IrreversibleMetastoreOperation(
+                                    format("drop partition %s.%s %s", schemaTableName, schemaTableName.getTableName(), partitionValues),
+                                    () -> delegate.dropPartition(schemaTableName.getSchemaName(), schemaTableName.getTableName(), partitionValues, true)));
+                            break;
+                        case ALTER: {
+                            String user = action.getUser();
+                            PartitionAndMore partitionAndMore = action.getData();
+                            Partition partition = partitionAndMore.getPartition();
+                            String targetLocation = partition.getStorage().getLocation();
+                            Optional<Partition> oldPartition = delegate.getPartition(schemaTableName.getSchemaName(), schemaTableName.getTableName(), partition.getValues());
+                            if (!oldPartition.isPresent()) {
+                                throw new PrestoException(
+                                        TRANSACTION_CONFLICT,
+                                        format("The partition you tried to modify was deleted by some one else. %s %s", schemaTableName.getTableName(), partition.getValues()));
+                            }
+                            String oldPartitionLocation = oldPartition.get().getStorage().getLocation();
+                            Path oldPartitionPath = new Path(oldPartitionLocation);
+
+                            // Location of the old partition and the new partition can be different because we allow arbitrary directories through LocationService.
+                            // If the location of the old partition is the same as the location of the new partition:
+                            // * Rename the old data directory to a temporary path with a special suffix
+                            // * Remember we will need to delete that directory at the end if transaction successfully commits
+                            // * Remember we will need to undo the rename if transaction aborts
+                            // Otherwise,
+                            // * Remember we will need to delete the location of the old partition at the end if transaction successfully commits
+                            if (targetLocation.equals(oldPartitionLocation)) {
+                                Path oldPartitionStagingPath = new Path(oldPartitionPath.getParent(), oldPartitionPath.getName() + "_" + action.getQueryId());
```

Comment:
Maybe add something like `temp` or `backup` to the name... also maybe start the directory name with an `_` so that is is considered "hidden"